File: lib/vision-safety.ts
/**
 * Google Cloud Vision API - SafeSearch
 * Merkmalsweise Schwellen (adult, violence, racy, medical, spoof).
 * Unterstützt:
 * - GOOGLE_CLOUD_VISION_API_KEY (API-Key)
 * - GOOGLE_VISION_PROJECT_ID + CLIENT_EMAIL + PRIVATE_KEY (Service Account, Fallback)
 *
 * DSGVO / Art. 44 DSGVO – Drittlandübermittlung:
 * Explizit EU-regionaler Endpunkt (eu-vision.googleapis.com).
 */
const VISION_API = "https://eu-vision.googleapis.com/v1/images:annotate"

export type SafeSearchLevel = "UNKNOWN" | "VERY_UNLIKELY" | "UNLIKELY" | "POSSIBLE" | "LIKELY" | "VERY_LIKELY"

export type SafeSearchCategory = "adult" | "violence" | "racy" | "medical" | "spoof"

/** Mindest-Likelihood für einen Verstoß, oder IGNORED (Merkmal wird nicht gewertet). */
export type CategoryViolationPolicy = SafeSearchLevel | "IGNORED"

const LEVEL_RANK: Record<string, number> = {
  UNKNOWN: 0,
  VERY_UNLIKELY: 1,
  UNLIKELY: 2,
  POSSIBLE: 3,
  LIKELY: 4,
  VERY_LIKELY: 5,
}

/** Google Vision Likelihood-Enum (protobuf) → Name */
const LIKELIHOOD_BY_NUMBER: Record<number, string> = {
  0: "UNKNOWN",
  1: "VERY_UNLIKELY",
  2: "UNLIKELY",
  3: "POSSIBLE",
  4: "LIKELY",
  5: "VERY_LIKELY",
}

/**
 * Produktions-Schwellen: Verstoß, wenn API-Wert ≥ konfigurierte Stufe.
 * adult: LIKELY+ · violence: POSSIBLE+ · racy: nur VERY_LIKELY · medical: LIKELY+ · spoof: ausgeschaltet
 */
export const SAFE_SEARCH_CATEGORY_POLICY: Record<SafeSearchCategory, CategoryViolationPolicy> = {
  adult: "LIKELY",
  violence: "POSSIBLE",
  racy: "VERY_LIKELY",
  medical: "LIKELY",
  spoof: "IGNORED",
}

interface SafeSearchAnnotation {
  adult?: SafeSearchLevel | string | number | unknown
  violence?: SafeSearchLevel | string | number | unknown
  racy?: SafeSearchLevel | string | number | unknown
  spoof?: SafeSearchLevel | string | number | unknown
  medical?: SafeSearchLevel | string | number | unknown
}

export interface SafetyCheckResult {
  safe: boolean
  reason?: string
  violation?: { key: string; level: string }
}

function normalizeLikelihood(level: unknown): string | null {
  if (level == null) return null
  if (typeof level === "string") return level
  if (typeof level === "number") return LIKELIHOOD_BY_NUMBER[level] ?? null
  const name = (level as { name?: string }).name
  if (typeof name === "string") return name
  return null
}

function isViolationForCategory(
  category: string,
  levelRaw: unknown,
  policy: Record<string, CategoryViolationPolicy>
): { violation: true; level: string } | null {
  const rule = policy[category]
  if (rule === undefined || rule === "IGNORED") return null
  const levelStr = normalizeLikelihood(levelRaw)
  if (!levelStr) return null
  const need = LEVEL_RANK[rule]
  const actual = LEVEL_RANK[levelStr]
  if (need === undefined || actual === undefined) return null
  if (actual >= need) return { violation: true, level: levelStr }
  return null
}

function checkAnnotation(
  annotation: SafeSearchAnnotation,
  policy: Record<string, CategoryViolationPolicy>
): SafetyCheckResult {
  for (const [key, levelRaw] of Object.entries(annotation)) {
    const cat = key.toLowerCase()
    const hit = isViolationForCategory(cat, levelRaw, policy)
    if (hit) {
      return {
        safe: false,
        reason: `Bild enthält möglicherweise ungeeignete Inhalte (${cat}).`,
        violation: { key: cat, level: hit.level },
      }
    }
  }
  return { safe: true }
}

export function hasVisionConfig(): boolean {
  const apiKey = process.env.GOOGLE_CLOUD_VISION_API_KEY?.trim()
  if (apiKey) return true
  const projectId = process.env.GOOGLE_VISION_PROJECT_ID?.trim()
  const clientEmail = process.env.GOOGLE_VISION_CLIENT_EMAIL?.trim()
  const privateKey = process.env.GOOGLE_VISION_PRIVATE_KEY?.trim()
  return !!(projectId && clientEmail && privateKey)
}

/** Für Health-Check: Welche Vision-Konfiguration ist aktiv (ohne Werte). */
export function getVisionConfigStatus(): { configured: boolean; method: "api_key" | "service_account" | null } {
  if (process.env.GOOGLE_CLOUD_VISION_API_KEY?.trim()) {
    return { configured: true, method: "api_key" }
  }
  const projectId = process.env.GOOGLE_VISION_PROJECT_ID?.trim()
  const clientEmail = process.env.GOOGLE_VISION_CLIENT_EMAIL?.trim()
  const privateKey = process.env.GOOGLE_VISION_PRIVATE_KEY?.trim()
  if (projectId && clientEmail && privateKey) {
    return { configured: true, method: "service_account" }
  }
  return { configured: false, method: null }
}

async function runSafeSearchWithClient(
  content: string,
  policy: Record<string, CategoryViolationPolicy>
): Promise<SafetyCheckResult> {
  const { ImageAnnotatorClient } = await import("@google-cloud/vision")
  const projectId = process.env.GOOGLE_VISION_PROJECT_ID!
  const clientEmail = process.env.GOOGLE_VISION_CLIENT_EMAIL!
  const privateKey = process.env.GOOGLE_VISION_PRIVATE_KEY!.replace(/\\n/g, "\n")
  const client = new ImageAnnotatorClient({
    credentials: { client_email: clientEmail, private_key: privateKey },
    projectId,
    apiEndpoint: "eu-vision.googleapis.com",
  })
  const [response] = await client.annotateImage({
    image: { content },
    features: [{ type: "SAFE_SEARCH_DETECTION" }],
  })
  const annotation = response.safeSearchAnnotation as SafeSearchAnnotation | undefined
  if (!annotation) return { safe: false, reason: "Bild konnte nicht geprüft werden.", violation: undefined }
  return checkAnnotation(annotation, policy)
}

async function runSafeSearchWithApiKey(
  content: string,
  policy: Record<string, CategoryViolationPolicy>
): Promise<SafetyCheckResult> {
  const apiKey = process.env.GOOGLE_CLOUD_VISION_API_KEY!
  const res = await fetch(`${VISION_API}?key=${apiKey}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      requests: [{
        image: { content },
        features: [{ type: "SAFE_SEARCH_DETECTION" }],
      }],
    }),
  })
  if (!res.ok) {
    console.error("[Vision] API error:", res.status, await res.text())
    return { safe: false, reason: "Bildprüfung fehlgeschlagen.", violation: undefined }
  }
  const data = await res.json()
  const ann = data?.responses?.[0]?.safeSearchAnnotation as Record<string, string> | undefined
  if (!ann) return { safe: false, reason: "Bild konnte nicht geprüft werden.", violation: undefined }
  const annotation: SafeSearchAnnotation = {
    adult: ann.adult,
    violence: ann.violence,
    racy: ann.racy,
    spoof: ann.spoof,
    medical: ann.medical,
  }
  return checkAnnotation(annotation, policy)
}

async function withTimeout<T>(promise: Promise<T>, ms: number, errMsg: string): Promise<T> {
  const timeout = new Promise<never>((_, reject) =>
    setTimeout(() => reject(new Error(errMsg)), ms)
  )
  return Promise.race([promise, timeout])
}

export async function checkImageSafety(buffer: Buffer): Promise<{ safe: boolean; reason?: string }> {
  if (!hasVisionConfig()) return { safe: false, reason: "Bildprüfung nicht konfiguriert." }

  const base64 = buffer.toString("base64")
  const run = () =>
    process.env.GOOGLE_CLOUD_VISION_API_KEY?.trim()
      ? runSafeSearchWithApiKey(base64, SAFE_SEARCH_CATEGORY_POLICY)
      : runSafeSearchWithClient(base64, SAFE_SEARCH_CATEGORY_POLICY)

  try {
    const result = await withTimeout(run(), 8000, "Timeout")
    return { safe: result.safe, reason: result.reason }
  } catch (err) {
    console.error("[Vision] Safety check failed:", err)
    return { safe: false, reason: "Bildprüfung fehlgeschlagen. Bitte versuche es erneut." }
  }
}

export type CheckImageSafetyFromBase64Options = {
  /** Überschreibt einzelne Merkmale (Rest bleibt SAFE_SEARCH_CATEGORY_POLICY). */
  policy?: Partial<Record<SafeSearchCategory, CategoryViolationPolicy>>
}

/** Pre-Check & Alert: Prüft Base64-Bild, gibt detailliertes Ergebnis (inkl. violation für Speicherung) */
export async function checkImageSafetyFromBase64(
  base64: string,
  options?: CheckImageSafetyFromBase64Options
): Promise<SafetyCheckResult> {
  if (!hasVisionConfig()) return { safe: false, reason: "Bildprüfung nicht konfiguriert.", violation: undefined }

  const policy: Record<string, CategoryViolationPolicy> = {
    ...SAFE_SEARCH_CATEGORY_POLICY,
    ...options?.policy,
  }
  const content = base64.replace(/^data:image\/\w+;base64,/, "").trim()
  const run = () =>
    process.env.GOOGLE_CLOUD_VISION_API_KEY?.trim()
      ? runSafeSearchWithApiKey(content, policy)
      : runSafeSearchWithClient(content, policy)

  try {
    return await withTimeout(run(), 8000, "Timeout")
  } catch (err) {
    console.error("[Vision] Safety check failed:", err)
    return { safe: false, reason: "Bildprüfung fehlgeschlagen.", violation: undefined }
  }
}

File: scripts/vision_safety.py
"""
Google Cloud Vision API – Safety Enforcement (Python).

- Nur SafeSearchDetection (keine Face Detection, keine Liveness/Spoof).
- Konfigurierbare Schwellenwerte und Blacklist.
- Nutzt die offizielle google-cloud-vision Library.
"""

import base64
import os
from typing import Optional

from google.cloud import vision
from google.cloud.vision_v1 import AnnotateImageResponse

from vision_config import (
    ENABLED_FEATURES,
    RESPONSE_BLACKLIST,
    SAFE_SEARCH_CATEGORIES,
    SAFE_SEARCH_THRESHOLDS,
)


def _get_client() -> Optional[vision.ImageAnnotatorClient]:
    """Erstellt den Vision-Client (Service Account oder ADC)."""
    project_id = os.environ.get("GOOGLE_VISION_PROJECT_ID")
    client_email = os.environ.get("GOOGLE_VISION_CLIENT_EMAIL")
    private_key = os.environ.get("GOOGLE_VISION_PRIVATE_KEY")

    if project_id and client_email and private_key:
        from google.oauth2 import service_account
        credentials = service_account.Credentials.from_service_account_info({
            "type": "service_account",
            "project_id": project_id,
            "client_email": client_email,
            "private_key": private_key.replace("\\n", "\n"),
        })
        return vision.ImageAnnotatorClient(
            credentials=credentials,
            client_options={"api_endpoint": "eu-vision.googleapis.com"},
        )

    api_key = os.environ.get("GOOGLE_CLOUD_VISION_API_KEY")
    if api_key:
        return vision.ImageAnnotatorClient(
            client_options={
                "api_endpoint": "eu-vision.googleapis.com",
                "api_key": api_key,
            }
        )

    return vision.ImageAnnotatorClient(
        client_options={"api_endpoint": "eu-vision.googleapis.com"}
    )


def _filter_response(response: AnnotateImageResponse) -> dict:
    """
    Entfernt blacklistete Response-Elemente.
    Nur safeSearchAnnotation wird für die Auswertung genutzt.
    """
    result = {}
    for key, value in response.__dict__.items():
        if key in RESPONSE_BLACKLIST or value is None:
            continue
        result[key] = value
    return result


def check_image_safety(
    image_content: bytes,
    *,
    thresholds: Optional[SafeSearchThresholds] = None,
) -> dict:
    """
    Prüft ein Bild auf SafeSearch-Verstöße.

    Args:
        image_content: Rohdaten des Bildes (Bytes).
        thresholds: Optionale Schwellenwerte (Standard: SAFE_SEARCH_THRESHOLDS).

    Returns:
        {
            "safe": bool,
            "reason": str | None,
            "violation": {"key": str, "level": str} | None,
        }
    """
    thresholds = thresholds or SAFE_SEARCH_THRESHOLDS
    client = _get_client()
    if not client:
        return {
            "safe": False,
            "reason": "Vision API nicht konfiguriert (fehlende ENV-Variablen).",
            "violation": None,
        }

    image = vision.Image(content=image_content)
    features = [vision.Feature(type_=f) for f in ENABLED_FEATURES]

    try:
        response = client.annotate_image(
            image=image,
            features=features,
        )
    except Exception as e:
        return {
            "safe": False,
            "reason": f"Bildprüfung fehlgeschlagen: {e}",
            "violation": None,
        }

    _filter_response(response)

    annotation = response.safe_search_annotation
    if not annotation:
        return {
            "safe": False,
            "reason": "Bild konnte nicht geprüft werden.",
            "violation": None,
        }

    for category in SAFE_SEARCH_CATEGORIES:
        level = getattr(annotation, category, None)
        if level is None:
            continue
        level_str = str(level.name) if hasattr(level, "name") else str(level)
        if thresholds.is_violation(category, level_str):
            return {
                "safe": False,
                "reason": f"Bild enthält möglicherweise ungeeignete Inhalte ({category}).",
                "violation": {"key": category, "level": level_str},
            }

    return {"safe": True, "reason": None, "violation": None}


def check_image_safety_base64(
    base64_content: str,
    *,
    thresholds: Optional[SafeSearchThresholds] = None,
) -> dict:
    """
    Prüft ein Base64-kodiertes Bild auf SafeSearch-Verstöße.

    Args:
        base64_content: Base64-String (mit oder ohne data-URL-Präfix).
        thresholds: Optionale Schwellenwerte.
    """
    content = base64_content.strip()
    if content.startswith("data:image/"):
        content = content.split(",", 1)[-1]
    try:
        image_bytes = base64.b64decode(content)
    except Exception as e:
        return {
            "safe": False,
            "reason": f"Ungültige Base64-Daten: {e}",
            "violation": None,
        }
    return check_image_safety(image_bytes, thresholds=thresholds)


if __name__ == "__main__":
    import sys
    if len(sys.argv) < 2:
        print("Usage: python vision_safety.py <path-to-image>")
        sys.exit(1)
    path = sys.argv[1]
    with open(path, "rb") as f:
        result = check_image_safety(f.read())
    print(result)
